extensions/devservices/keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/DevServicesProcessor.java

```diff
@@ -0,0 +1,398 @@
+import io.quarkus.deployment.builditem.ServiceStartBuildItem;
+import io.quarkus.devservices.keycloak.runtime.DevServicesConfig;
+import io.quarkus.devservices.keycloak.runtime.KeycloakBuildTimeConfig;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ConfigUtils;
+
+public class DevServicesProcessor {
+    private static final Logger LOG = Logger.getLogger(DevServicesProcessor.class);
+
+    private static final String CONFIG_PREFIX = "quarkus.oidc.";
+    private static final String AUTH_SERVER_URL_CONFIG_KEY = CONFIG_PREFIX + "auth-server-url";
+    private static final String APPLICATION_TYPE_CONFIG_KEY = CONFIG_PREFIX + "application-type";
+    private static final String CLIENT_ID_CONFIG_KEY = CONFIG_PREFIX + "client-id";
+    private static final String CLIENT_SECRET_CONFIG_KEY = CONFIG_PREFIX + "credentials.secret";
+    private static final String KEYCLOAK_URL_KEY = "keycloak.url";
+
+    private static final String KEYCLOAK_VERSION = System.getProperty("keycloak.version", "latest");
+    private static final String KEYCLOAK_IMAGE_QUAY_REP = "quay.io/keycloak/keycloak:";
+    private static final String KEYCLOAK_IMAGE_QUAY_NAME = KEYCLOAK_IMAGE_QUAY_REP + KEYCLOAK_VERSION;
+
+    private static final int KEYCLOAK_EXPOSED_PORT = 8080;
+    private static final String KEYCLOAK_DOCKER_REALM_PATH = "/tmp/realm.json";
+    private static final String KEYCLOAK_USER_PROP = "KEYCLOAK_USER";
+    private static final String KEYCLOAK_PASSWORD_PROP = "KEYCLOAK_PASSWORD";
+    private static final String KEYCLOAK_VENDOR_PROP = "DB_VENDOR";
+    private static final String KEYCLOAK_IMPORT_PROP = "KEYCLOAK_IMPORT";
+    private static final String KEYCLOAK_ADMIN_USER = "admin";
+    private static final String KEYCLOAK_ADMIN_PASSWORD = "admin";
+    private static final String KEYCLOAK_DB_VENDOR = "H2";
+    private static volatile List<Closeable> closeables;
+    private static volatile DevServicesConfig capturedDevServicesConfiguration;
+    private static volatile boolean first = true;
+    private final IsDockerWorking isDockerWorking = new IsDockerWorking(true);
+
+    KeycloakBuildTimeConfig config;
+
+    @BuildStep(onlyIfNot = IsNormal.class)
+    public DevServicesConfigBuildItem startOIDCContainer(LaunchModeBuildItem launchMode,
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
+            BuildProducer<ServiceStartBuildItem> serviceStartBuildItemBuildProducer,
+            BuildProducer<DevServicesNativeConfigResultBuildItem> devServices) {
+
+        DevServicesConfig currentDevServicesConfiguration = config.devservices;
+        // figure out if we need to shut down and restart any existing Keycloak container
+        // if not and the Keycloak container have already started we just return
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.TEST;
```

Comment:
This should not be restarted for every test run. Looks like some other extensions do this but this is likely due to a potential 'residual data' issue which does not affect keycloak. These other extensions that do this likely need to be fixed as this will significantly slow down continuous testing.